src/app/lifecycle.ts

```diff
@@ -15,30 +15,116 @@
+    if (typeof options === 'undefined') {
+      options = loadOptionsFromEnvVar();
+      options.credential = getApplicationDefault();
+    }
+
+    if (typeof appName !== 'string' || appName === '') {
+      throw new FirebaseAppError(
+        AppErrorCodes.INVALID_APP_NAME,
+        `Invalid Firebase app name "${appName}" provided. App name must be a non-empty string.`,
+      );
+    } else if (this.appStore.has(appName)) {
+      if (appName === DEFAULT_APP_NAME) {
+        throw new FirebaseAppError(
+          AppErrorCodes.DUPLICATE_APP,
+          'The default Firebase app already exists. This means you called initializeApp() ' +
+          'more than once without providing an app name as the second argument. In most cases ' +
+          'you only need to call initializeApp() once. But if you do want to initialize ' +
+          'multiple apps, pass a second argument to initializeApp() to give each app a unique ' +
+          'name.',
+        );
+      } else {
+        throw new FirebaseAppError(
+          AppErrorCodes.DUPLICATE_APP,
+          `Firebase app named "${appName}" already exists. This means you called initializeApp() ` +
+          'more than once with the same app name as the second argument. Make sure you provide a ' +
+          'unique name every time you call initializeApp().',
+        );
+      }
+    }
+
+    const app = new FirebaseApp(options, appName, this);
+    this.appStore.set(app.name, app);
+    return app;
+  }
+
+  public getApp(appName: string = DEFAULT_APP_NAME): App {
+    if (typeof appName !== 'string' || appName === '') {
+      throw new FirebaseAppError(
+        AppErrorCodes.INVALID_APP_NAME,
+        `Invalid Firebase app name "${appName}" provided. App name must be a non-empty string.`,
+      );
+    } else if (!this.appStore.has(appName)) {
+      let errorMessage: string = (appName === DEFAULT_APP_NAME)
+        ? 'The default Firebase app does not exist. ' : `Firebase app named "${appName}" does not exist. `;
+      errorMessage += 'Make sure you call initializeApp() before using any of the Firebase services.';
+
+      throw new FirebaseAppError(AppErrorCodes.NO_APP, errorMessage);
+    }
+
+    return this.appStore.get(appName)!;
+  }
+
+  public getApps(): App[] {
+    // Return a copy so the caller cannot mutate the array
+    return Array.from(this.appStore.values());
+  }
+
+  public deleteApp(app: App): Promise<void> {
+    if (typeof app !== 'object' || app === null || !('options' in app)) {
+      throw new FirebaseAppError(AppErrorCodes.INVALID_ARGUMENT, 'Invalid app argument.');
+    }
+
+    // Make sure the given app already exists.
+    const existingApp = getApp(app.name);
 
-export function initializeApp(options?: AppOptions, name?: string): App {
-  return defaultNamespace.initializeApp(options, name);
+    // Delegate delete operation to the App instance itself.
+    return (existingApp as FirebaseApp).delete();
```

Comment:
Should the app be removed from `appStore`?

Reply:
We do that in a somewhat roundabout way:

`FirebaseApp.delete()` --> `AppStore.removeApp()`

This is needed because we (for the moment) have 2 ways to delete an App instance. See my other comment.